internal/client/connection.go

```diff
@@ -38,7 +38,195 @@ const (
+		HTTP:    http.StatusNotFound,
+		CoAP:    codes.NotFound,
+		Name:    "not found",
+		Success: false,
+	}
+	CodeMethodNotAllowed = ResponseCode{
+		HTTP:    http.StatusMethodNotAllowed,
+		CoAP:    codes.MethodNotAllowed,
+		Name:    "method not allowed",
+		Success: false,
+	}
+	CodeNotAcceptable = ResponseCode{
+		HTTP:    http.StatusNotAcceptable,
+		CoAP:    codes.NotAcceptable,
+		Name:    "not acceptable",
+		Success: false,
+	}
+	CodePreconditionFailed = ResponseCode{
+		HTTP:    http.StatusPreconditionFailed,
+		CoAP:    codes.PreconditionFailed,
+		Name:    "precondition failed",
+		Success: false,
+	}
+	CodeRequestEntityTooLarge = ResponseCode{
+		HTTP:    http.StatusRequestEntityTooLarge,
+		CoAP:    codes.RequestEntityTooLarge,
+		Name:    "request entity too large",
+		Success: false,
+	}
+	CodeUnsupportedContentFormat = ResponseCode{
+		HTTP:    http.StatusUnsupportedMediaType,
+		CoAP:    codes.UnsupportedMediaType,
+		Name:    "unsupported content format",
+		Success: false,
+	}
+	// Server error codes
+	// https://tools.ietf.org/html/rfc7252#section-5.9.3
+	CodeInternalServerError = ResponseCode{
+		HTTP:    http.StatusInternalServerError,
+		CoAP:    codes.InternalServerError,
+		Name:    "internal server error",
+		Success: false,
+	}
+	CodeNotImplemented = ResponseCode{
+		HTTP:    http.StatusNotImplemented,
+		CoAP:    codes.NotImplemented,
+		Name:    "not implemented",
+		Success: false,
+	}
+	CodeBadGateway = ResponseCode{
+		HTTP:    http.StatusBadGateway,
+		CoAP:    codes.BadGateway,
+		Name:    "bad gateway",
+		Success: false,
+	}
+	CodeServiceUnavailable = ResponseCode{
+		HTTP:    http.StatusServiceUnavailable,
+		CoAP:    codes.ServiceUnavailable,
+		Name:    "service unavailable",
+		Success: false,
+	}
+	CodeGatewayTimeout = ResponseCode{
+		HTTP:    http.StatusGatewayTimeout,
+		CoAP:    codes.GatewayTimeout,
+		Name:    "gateway timeout",
+		Success: false,
+	}
+	CodeProxyingNotSupported = ResponseCode{
+		HTTP:    0, // no mapping
+		CoAP:    codes.ProxyingNotSupported,
+		Name:    "proxying not supported",
+		Success: false,
+	}
+)
+
+// ResponseCodes list all the mapped response codes
+var ResponseCodes = []ResponseCode{
+	CodeCreated,
+	CodeDeleted,
+	CodeValid,
+	CodeChanged,
+	CodeContent,
+	CodeBadRequest,
+	CodeUnauthorized,
+	CodeBadOption,
+	CodeForbidden,
+	CodeNotFound,
+	CodeMethodNotAllowed,
+	CodeNotAcceptable,
+	CodePreconditionFailed,
+	CodeRequestEntityTooLarge,
+	CodeUnsupportedContentFormat,
+	CodeInternalServerError,
+	CodeNotImplemented,
+	CodeBadGateway,
+	CodeServiceUnavailable,
+	CodeGatewayTimeout,
+	CodeProxyingNotSupported,
+}
+
+// ResponseCode is used to relay the outcome of HTTP/CoAP requests made to AM/Gateway
+type ResponseCode struct {
+	HTTP    int
+	CoAP    codes.Code
+	Name    string
+	Success bool
+}
+
+// IsWrappedIn will check if the given error is a ResponseError and if it wraps this ResponseCode
+func (r ResponseCode) IsWrappedIn(err error) bool {
```

Comment:
If ResponseCode implemented the `Unwrap` interface then we could use the inbuilt 'Is' and 'As' error wrapping functionality available from 1.13 instead of this custom method.

Reply:
I didn't really want to turn `ResponseCode` into an `error`, because not all codes are errors. I think `Unwrap` is specifically for errors wrapped in errors.

Reply:
My mistake, I merged `ResponseCode` and `ResponseError` into a single struct in my head.